flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/fanout/FanOutShardSubscriber.java

```diff
@@ -0,0 +1,466 @@
+			this.waitForSubscriptionLatch = waitForSubscriptionLatch;
+		}
+
+		/**
+		 * Flag to the producer that we are ready to receive more events.
+		 */
+		public void requestRecord() {
+			if (!cancelled) {
+				LOG.debug("Requesting more records from EFO subscription - {} ({})", shardId, consumerArn);
+				subscription.request(1);
+			}
+		}
+
+		@Override
+		public void onSubscribe(Subscription subscription) {
+			this.subscription = subscription;
+			waitForSubscriptionLatch.countDown();
+		}
+
+		@Override
+		public void onNext(SubscribeToShardEventStream subscribeToShardEventStream) {
+			subscribeToShardEventStream.accept(new SubscribeToShardResponseHandler.Visitor() {
+				@Override
+				public void visit(SubscribeToShardEvent event) {
+					synchronized (lockObject) {
+						if (enqueueEventWithRetry(new SubscriptionNextEvent(event))) {
+							requestRecord();
+						}
+					}
+				}
+			});
+		}
+
+		@Override
+		public void onError(Throwable throwable) {
+			LOG.debug("Error occurred on EFO subscription: {} - ({}).  {} ({})",
+				throwable.getClass().getName(), throwable.getMessage(), shardId, consumerArn);
+
+			// Cancel the subscription to signal the onNext to stop queuing and requesting data
+			cancelSubscription();
+
+			synchronized (lockObject) {
+				// Empty the queue and add a poison pill to terminate this subscriber
+				// The synchronized block ensures that new data is not written in the meantime
+				queue.clear();
+				enqueueEvent(new SubscriptionErrorEvent(throwable));
+			}
+		}
+
+		@Override
+		public void onComplete() {
+			LOG.debug("EFO subscription complete - {} ({})", shardId, consumerArn);
+			enqueueEvent(new SubscriptionCompleteEvent());
+		}
+
+		private void cancelSubscription() {
+			if (!cancelled) {
+				cancelled = true;
+				subscription.cancel();
+			}
+		}
+
+		/**
+		 * Continuously attempt to enqueue an event until successful or the subscription is cancelled (due to error).
+		 * When backpressure applied by the consumer exceeds 30s for a single batch, a ReadTimeoutException will be
+		 * thrown by the network stack. This will result in the subscription be cancelled and this event being discarded.
+		 * The subscription would subsequently be reacquired and the discarded data would be fetched again.
+		 *
+		 * @param event the event to enqueue
+		 * @return true if the event was successfully enqueued.
+		 */
+		private boolean enqueueEventWithRetry(final FanOutSubscriptionEvent event) {
+			boolean result = false;
+			do {
+				if (cancelled) {
+					break;
+				}
+
+				synchronized (lockObject) {
+					result = enqueueEvent(event);
+				}
+			} while (!result);
+
+			return result;
+		}
+
+		/**
+		 * Offers the event to the queue.
+		 *
+		 * @param event the event to enqueue
+		 * @return true if the event was successfully enqueued.
+		 */
+		private boolean enqueueEvent(final FanOutSubscriptionEvent event) {
+			try {
+				if (!queue.offer(event, ENQUEUE_WAIT_SECONDS, SECONDS)) {
```

Comment:
Yes correct. This is a measure to prevent lock up in high back-pressure scenarios. If the app is applying large back-pressure and `onNext` is blocked on queue push, this 5 second retry allows `onError` to set `cancel = true` and then slip the error event into the queue. The downside to this is that we might end up throwing away 2 record batches and re-requesting them on next subscription. However my testing revealed that the AWS SDK Async client and Netty client do not like these threads hanging around and it is happier this way.